### ISSUE TITLE
Consider killing `@inner` in favor of regular render

### ISSUE THREAD

**Issue by chrismccord** [phoenixframework/phoenix#1358]
For 1.1, we should consider killing `@inner` as it's not needed. If we expose `template_name` or similar, we can do this in the layout:

    <%= render view_module(@conn), template_name(@conn), assigns %>

This would change the render order, which shouldn't matter, and it would remove one unclear aspect of layouts (the `@inner` assign). Thoughts?

**Comment by josevalim** [comment]
Sounds good!

On Wednesday, November 18, 2015, Chris McCord <notifications@github.com>
wrote:

> For 1.1, we should consider killing @inner as it's not needed. If we
> expose template_name or similar, we can do this in the layout:
>
> <%= render view_module(@conn), template_name(@conn), assigns %>
>
> This would change the render order, which shouldn't matter, and it would
> remove one unclear aspect of layouts (the @inner assign). Thoughts?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/phoenixframework/phoenix/issues/1358>.
>


-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Director of R&D

**Comment by Gazler** [comment]
:+1: The `@inner' variable has always felt a little weird. It's not like `render` which you can look up in the docs.

**Comment by stevedomin** [comment]
:+1: 

@chrismccord: keen to tackle that one if you're not on it already :)

**Comment by chrismccord** [comment]
Thanks @stevedomin , but I'm on it :)

**Comment by bryansray** [comment]
:thumbsup:

Would it be possible to add a more terse method? Unfortunately, I don't know enough about the language or framework to say ... but 

    <%= render view_module(@conn), template_name(@conn), assigns %>

feels a bit long-winded.

**Comment by chrismccord** [comment]
> Would it be possible to add a more terse method? Unfortunately, I don't know enough about the language or framework to say ... but

If it helps, this line would be generated for you in new projects in the default layout, but I am :-1: on a more terse function, exactly because it shows that there's nothing special going on. Templates are functions, and we can render other templates by calling functions on other view modules. So the slightly expanded form is preferable.

**Comment by bryansray** [comment]
I see your point, but are you not sacrificing readability of the code? 

As a new-comer to a framework, typically layouts should feel like childs play. You can see this in a lot of frameworks: rails, MVC.NET, express, are ones that I have experience with, but ... when I come to phoenix and suddenly have to decipher:

  <%= render view_module(@conn), template_name(@conn), assigns %>

just to render a main view. I mean almost _nothing_ about that line of code really helps me figure out what is going on.

**Comment by bryansray** [comment]
Posting this again, because I don't think emails come through after edits.

Again, let me ask this question: is it possible that I could provide my own function to simplify that at all? It doesn't necessary have to be provided by the framework (although, i think it would be a good idea).

**Comment by chrismccord** [comment]
The takeaway for me is we have to compare what we have now:

    <%= @inner %>

vs

    <%= render view_module(@conn), template_name(@conn), assigns %>

The former is magically available. The latter users three functions, all of which are documented and can be looked up.

> is it possible that I could provide my own function to simplify that at all? It doesn't necessary have to be provided by the framework (although, i think it would be a good idea).

Of course! It's functions all the way down :)

```elixir
# <%= render_inner(conn, assigns) %>
def render_inner(conn, assigns) do
  render(view_module(conn), template_name(conn), assigns)
end
```

**Comment by bryansray** [comment]
:+1: 

Awesome. Thanks, @chrismccord!

**Comment by paulcsmith** [comment]
Not sure if this should be a separate issue or not, but the `assigns` part of this was pretty confusing to me. Where does the `assigns` function come from? I don't see anything in the docs and I don't have any ideas how it works. Maybe that could be changed at the same time so that it's a bit more clear? 

Would this work?

`<%= render view_module(@conn), template_name(@conn), @conn.assigns %>`? If not, I think it *should* since it makes a lot more sense, but maybe that's due to my ignorance of the Phoenix stack.

**Comment by Gazler** [comment]
There was a discussion on IRC about using @conn. I am on my phone, but here
is the link.

https://botbot.me/freenode/elixir-lang/2015-11-18/?msg=54438350&page=15

**Comment by chrismccord** [comment]
@paulcsmith `assigns` comes directly from EEx (http://elixir-lang.org/docs/v1.0/eex/)

> The assigns extension is useful when the number of variables required by the template is not specified at compilation time.

`@conn.assigns` should *not* be used, because it couples your template to the connection. We already merge all conn.assigns into the EEx assigns for you, so adding the `@conn.assigns` gains you nothing but coupling. I do think we could cover `assigns` better in the view/template docs somewhere. PR?

**Comment by paulcsmith** [comment]
@chrismccord I'm pretty swamped, but I can open a PR when I have some free time for sure! Thanks for explaining everything. I think merging in the assigns automatically is a bit too magical, but I see your point about not wanting to couple things too tightly. Maybe we can think of a better solution later. For now I'll focus on improving the docs when I have some more time!

**Comment by chrismccord** [comment]
> I think merging in the assigns automatically is a bit too magical, but I see your point about not wanting to couple things too tightly. Maybe we can think of a better solution later.

I'll wait on the PR, but I remain convinced the current solution is best. Thanks!

**Comment by chrismccord** [comment]
Closed via https://github.com/phoenixframework/phoenix/commit/3f5062872cfe8f606e180edc5295b50693692b06
